contracts/account/paymaster/PaymasterCore.sol

```diff
@@ -0,0 +1,133 @@
+ *
+ * The paymaster includes support to call the {IEntryPointStake} interface to manage the paymaster's deposits and stakes
+ * through the internal functions {_deposit}, {_withdraw}, {_addStake}, {_unlockStake} and {_withdrawStake}.
+ *
+ * * Deposits are used to pay for user operations.
+ * * Stakes are used to guarantee the paymaster's reputation and obtain more flexibility in accessing storage.
+ *
+ * NOTE: See [Paymaster's unstaked reputation rules](https://eips.ethereum.org/EIPS/eip-7562#unstaked-paymasters-reputation-rules)
+ * for more details on the paymaster's storage access limitations.
+ */
+abstract contract PaymasterCore is IPaymaster {
+    /// @dev Unauthorized call to the account.
+    error PaymasterUnauthorized(address sender);
+
+    /// @dev Revert if the caller is not the entry point.
+    modifier onlyEntryPoint() {
+        _checkEntryPoint();
+        _;
+    }
+
+    /// @dev Canonical entry point for the account that forwards and validates user operations.
+    function entryPoint() public view virtual returns (IEntryPoint) {
+        return IEntryPoint(0x0000000071727De22E5E9d8BAf0edAc6f37da032);
+    }
+
+    /// @inheritdoc IPaymaster
+    function validatePaymasterUserOp(
+        PackedUserOperation calldata userOp,
+        bytes32 userOpHash,
+        uint256 maxCost
+    ) public virtual onlyEntryPoint returns (bytes memory context, uint256 validationData) {
+        return _validatePaymasterUserOp(userOp, userOpHash, maxCost);
+    }
+
+    /**
+     * @dev Internal validation of whether the paymaster is willing to pay for the user operation.
+     * Returns the context to be passed to postOp and the validation data.
+     *
+     * The `requiredPreFund` is the amount the paymaster has to pay (in native tokens). It's calculated
+     * as `requiredGas * userOp.maxFeePerGas`, where `required` gas can be calculated from the user operation
+     * as `verificationGasLimit + callGasLimit + paymasterVerificationGasLimit + paymasterPostOpGasLimit + preVerificationGas`
+     */
+    function _validatePaymasterUserOp(
+        PackedUserOperation calldata userOp,
+        bytes32 userOpHash,
+        uint256 requiredPreFund
+    ) internal virtual returns (bytes memory context, uint256 validationData);
+
+    /// @inheritdoc IPaymaster
+    function postOp(
+        PostOpMode mode,
+        bytes calldata context,
+        uint256 actualGasCost,
+        uint256 actualUserOpFeePerGas
+    ) public virtual onlyEntryPoint {
+        _postOp(mode, context, actualGasCost, actualUserOpFeePerGas);
+    }
+
+    /**
+     * @dev Handles post user operation execution logic. The caller must be the entry point.
+     *
+     * It receives the `context` returned by `_validatePaymasterUserOp`. Reverts by default
+     * since the function is not called if no context is returned by {validatePaymasterUserOp}.
+     * Overriding without calling `super._postOp(...)` is advised.
```

Comment:
I'd like to inderstand why reverting is necessary. My initial idea was that this would just be a no-op, ignoring any unexpected context.

Are there any security concern with doing nothing here? I'm concerned that advising to not call super would cause Paymaster extensions to become incompatible.

Reply:
The thing is that the entrypoint will call `postOp` when there's context but the user op won't revert if the `postOp` reverted. We have 2 options:

1) Revert by default, users who want to implement `validateUserOp` won't have an issue unless they really want to implement `postOp`, where overriding without calling super is ok.
2) Leave `_postOp` virtual, requiring that everyone implements it without knowing that it can just revert and it will be processed by the entrypoint regardless. I see very tempting that users will leave the function empty, which I think is even worse because in those cases they're indeed not calling `super`

So my reasoning is that 1) is safer. wdyt?

Reply:
What about:

3. do nothing by default. Users can do whatever they want. Returning an empty context would mean the `postOp` is not called, which is ok. Returning a non empty context would mean the `postOp` is called, and does nothing, whuch is also ok. Users that want to work with a context can return it ... and override `_postOp`. When doing so, they can call super (recommandend good practice when overriding) and that would not revert.

Reply:
[A bit like the `_afterTokenTransfer` hook in v4](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.9.6/contracts/token/ERC20/ERC20.sol#L350-L364)

Reply:
I felt 3 would be dangerous too. If the linearization reaches the last `_postOp` that doesn't call super, it may be missing another override at the same hierarchic level. (e.g. `contract A is B, PaymasterCore {}` where B implements `_postOp`. I think B would be skipped).

I'm keeping it empty for now, I don't think such situation is common

Reply:
In option 3, if one of the override doesn't call super, then you are definitelly "losing" some logic (just like the transfer hooks) ... but it would be clearly recommanded that you should call super.

In option 1, there is no super call, so you are certain that logic is dropped is you don't manually linearise the calls. IMO that is the most tricky/error prone approach.